runtime/memprof.c

```diff
@@ -77,55 +76,108 @@ static intnat callstack_buffer_len = 0;
-      y = (mt_state[i] & 0x80000000) + (mt_state[i+1] & 0x7fffffff);
-      mt_state[i] = mt_state[i+397] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    }
-    for (i = 227; i < MT_STATE_SIZE - 1; i++) {
-      y = (mt_state[i] & 0x80000000) + (mt_state[i+1] & 0x7fffffff);
-      mt_state[i] = mt_state[i-227] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    }
-    y = (mt_state[MT_STATE_SIZE - 1] & 0x80000000) + (mt_state[0] & 0x7fffffff);
-    mt_state[MT_STATE_SIZE - 1] =
-      mt_state[396] ^ (y >> 1) ^ ((-(y&1)) & 0x9908b0df);
-    mt_index = 0;
+  /* Instead of using temporary buffers, we could use one big loop,
+     but it turns out SIMD optimizations of compilers are more fragile
+     when using larger loops.  */
+  static uint32_t buff32[MT_STATE_SIZE];
+  static float bufff[MT_STATE_SIZE];
+
+  CAMLassert(lambda > 0.);
+
+  /* Shuffle the Mersenne Twister state */
+  for (i = 0; i < 227; i++)
+    mt_state[i] = mt_shuffle(mt_state[i], mt_state[i+1], mt_state[i+397]);
+  for (i = 227; i < MT_STATE_SIZE - 1; i++)
+    mt_state[i] = mt_shuffle(mt_state[i], mt_state[i+1], mt_state[i-227]);
+  mt_state[MT_STATE_SIZE - 1] =
+    mt_shuffle(mt_state[MT_STATE_SIZE - 1], mt_state[0], mt_state[396]);
+
+  /* Generate 32bits random variables */
+  for(i = 0; i < MT_STATE_SIZE; i++) {
+    uint32_t y = mt_state[i];
+    y = y ^ (y >> 11);
+    y = y ^ ((y << 7) & 0x9d2c5680);
+    y = y ^ ((y << 15) & 0xefc60000);
+    y = y ^ (y >> 18);
+    buff32[i] = y;
   }
 
-  y = mt_state[mt_index];
-  y = y ^ (y >> 11);
-  y = y ^ ((y << 7) & 0x9d2c5680);
-  y = y ^ ((y << 15) & 0xefc60000);
-  y = y ^ (y >> 18);
+  /* Generate exponential random variables by computing logarithms. We
+     do not use math.h library functions, which are slow and prevent
+     compiler from using SIMD instructions. */
+  for(i = 0; i < MT_STATE_SIZE; i++) {
+    union { float f; int32_t i; } u;
+    float exp, x;
+    u.f = buff32[i] + 0.5f; /* We convert the uniform variable to a float ... */
+    exp = (u.i >> 23) - 162;/* ... of which we extract the exponent ... */
+    u.i = (u.i & 0x7FFFFF) | 0x3F800000;
+    x = u.f;                /* ... and the mantissa. */
+
+    x =
+      /* This polynomial computes the logarithm of the mantissa (which
+         is in [1, 2]) up to a relatively good precision (< 1.5e-5). */
+      x * (3.539886362f + x * (-2.473745096f + x * (1.137731547f +
+        x * (-0.2906512007f + x *  3.129494483e-2f))))
+
+      /* We add the term corresponding to the exponent, and additive
+         constants.
+
+         0.13492425 =
+            - 1.94451655   (the constant term of the polynomial)
+            + 3*log(2)     (compensating the shifting of the exponent)
+            - 7.5e-7       (subtract a small constant to make sure that x < 0
+                            despite rounding errors)
+      */
+      + (0.13492425f + 0.6931471805f*exp);
+
+    /* => x is (up to a relatively good precision) log(unif[i] + 0.5)/2^32. */
```

Comment:
Would it make sense to turn this log computation into an inline function, just like you did for `mt_shuffle`?  Vectorization would still work, based on a quick experiment with GCC.